### DISCUSSION TITLE
Use of zigpy database

### DISCUSSION THREAD

**Discussion by pipiche38** [zigpy/zigpy#1361] [Q&A]
@puddly I have started investigating the use of the zigpy persistent Db and also going up in the stack.

for now _load_db() works well, and I can restart with a save state.

Where I have issues, is that the list of devices, which are save seems to be random (or somehow )

If I do a query on the Db on devices_v12 I find the coordnator and 1 device (router) out of 4 others which have been paired.

My question, is more what is your guidance to debug and understand where the issue is.

From the log I saw that the device_initialisation() is started with all of the discovery, but what is missing to get the device store in the zigpy DB ?

**Comment by Hedda** [comment]
> I have started investigating the use of the zigpy persistent Db and also going up in the stack.

@pipiche38 off-topic; but if you do decide to change your [ZigBee For Domoticz Plugin](https://github.com/zigbeefordomoticz/Domoticz-Zigbee) to make use of high-level functions of zigpy, any chance that you will then consider further contributions to help improve zigpy's development documentation (i.e. zigpy API and examples)?

https://github.com/zigpy/zigpy/pull/477

https://github.com/zigpy/zigpy/issues/471

Request for more development documentation is still open issues in the main zigpy repository and radio libraries like bellows and zigpy-znp:

https://github.com/zigpy/zigpy/issues/1087

https://github.com/zigpy/bellows/issues/431

https://github.com/zigpy/zigpy-znp/issues/192

zigpy's wiki and CONTRIBUTING.md file currently only contain snippets of references which does not make it easy for those new to zigpy:

https://github.com/zigpy/zigpy/wiki

https://github.com/zigpy/zigpy/blob/dev/CONTRIBUTING.md#the-zigpy-api

**Comment by pipiche38** [comment]
Progressing, but not yet getting updates on the persistent Db

for example:

2024-03-09 16:43:20,061 INFO    :b"packet_received - AddrModeAddress(addr_mode=<AddrMode.NWK: 2>, address=0x9C64) 260 6 1 1 b'\\x18\\x16\\n\\x00\\x00\\x10\\x01' 18160a00001001 AddrMode.NWK"
2024-03-09 16:43:20,068 INFO    :b'Known device <Device model=None manuf=None nwk=0x9C64 ieee=00:04:74:00:00:1f:09:a4 is_initialized=True>'

which is mainly reached after starting `device.schedule_initialize()` when the device is not initialized

So I guess something is not launching the write to the PersistingListener, despite the fact that it is launched at startup when loading the Db

```
async def _load_db(self) -> None:
    LOGGER.info("_load_db")
    database_file = self.config[zigpy_conf.CONF_DATABASE]
    if not database_file:
        return

    LOGGER.info("PersistingListener on %s" %database_file)
    self._dblistener = await zigpy.appdb.PersistingListener.new(database_file, self)
    await self._dblistener.load()
    self._add_db_listeners()


def _add_db_listeners(self):
    LOGGER.info("_add_db_listeners")
    if self._dblistener is None:
        return

    self.add_listener(self._dblistener)
    self.groups.add_listener(self._dblistener)
    self.backups.add_listener(self._dblistener)
    self.topology.add_listener(self._dblistener)


```

**Reply by pipiche38** [reply]
for now, I found that if I received a packet from an Initialized device, I can do `self.device_initialized( device )` and then it is store in the PersistentListener

Most likely there is something that I missed
